internal/app/machined/pkg/runtime/v1alpha1/v1alpha1_events.go

```diff
@@ -5,97 +5,122 @@
+	mu sync.Mutex
+	c  *sync.Cond
 }
 
 // NewEvents initializes and returns the v1alpha1 runtime event stream.
-func NewEvents(n int) *Events {
+func NewEvents(cap int) *Events {
 	e := &Events{
-		subscribers: make([]chan machine.Event, 0, n),
+		stream: make([]runtime.Event, cap),
+		cap:    cap,
 	}
 
+	e.c = sync.NewCond(&e.mu)
+
 	return e
 }
 
 // Watch implements the Events interface.
 func (e *Events) Watch(f runtime.WatchFunc) {
-	ch := e.add()
+	ctx, ctxCancel := context.WithCancel(context.Background())
+
+	ch := make(chan runtime.Event)
 
 	go func() {
-		defer close(ch)
-		defer e.delete(ch)
+		defer ctxCancel()
 
 		f(ch)
 	}()
-}
 
-// Publish implements the Events interface.
-func (e *Events) Publish(msg proto.Message) {
-	value, err := proto.Marshal(msg)
-	if err != nil {
-		log.Printf("failed to marshal message: %v", err)
-
-		return
-	}
-
-	event := machine.Event{
-		Data: &any.Any{
-			// In the future, we can publish `talos/runtime`, and
-			// `talos/plugin/<plugin>` (or something along those lines) events.
-			TypeUrl: fmt.Sprintf("talos/runtime/%s", proto.MessageName(msg)),
-			Value:   value,
-		},
-	}
+	e.mu.Lock()
+	pos := e.writePos
+	gen := e.gen
+	e.mu.Unlock()
 
-	e.Lock()
-	defer e.Unlock()
+	go func() {
+		defer close(ch)
 
-	for _, sub := range e.subscribers {
-		// drop the event if some subscriber is stuck
-		// dropping is bad, but better than blocking event propagation
-		select {
-		case sub <- event:
-		default:
+		for {
+			e.mu.Lock()
+			for pos == e.writePos {
+				e.c.Wait()
+
+				select {
+				case <-ctx.Done():
+					e.mu.Unlock()
+					return
+				default:
+				}
+			}
+
+			if e.gen > gen && pos < e.writePos {
+				// buffer overrun, there's no way to signal error in this case,
+				// so for now just return
+				e.mu.Unlock()
+				return
+			}
+
+			event := e.stream[pos]
+			pos = (pos + 1) % e.cap
+
+			if pos == 0 {
+				gen = e.gen
+			}
+
+			e.mu.Unlock()
+
+			select {
+			case ch <- event:
+			case <-ctx.Done():
+				return
+			}
 		}
-	}
+	}()
 }
 
-func (e *Events) add() chan machine.Event {
-	e.Lock()
-	defer e.Unlock()
-
-	ch := make(chan machine.Event, 100)
-
-	e.subscribers = append(e.subscribers, ch)
+// Publish implements the Events interface.
+func (e *Events) Publish(msg proto.Message) {
+	event := runtime.Event{
+		// In the future, we can publish `talos/runtime`, and
+		// `talos/plugin/<plugin>` (or something along those lines) events.
+		TypeURL: fmt.Sprintf("talos/runtime/%s", proto.MessageName(msg)),
+		Payload: msg,
+	}
 
-	return ch
-}
+	e.mu.Lock()
+	defer e.mu.Unlock()
 
-func (e *Events) delete(ch chan machine.Event) {
-	e.Lock()
-	defer e.Unlock()
+	e.stream[e.writePos] = event
+	e.writePos = (e.writePos + 1) % e.cap
 
-	for i, sub := range e.subscribers {
-		if sub == ch {
-			l := len(e.subscribers)
-			e.subscribers[i] = e.subscribers[l-1]
-			e.subscribers[l-1] = nil
-			e.subscribers = e.subscribers[:l-1]
-		}
+	if e.writePos == 0 {
+		e.gen++
```

Comment:
IIUC we should never have a generation of zero?

Reply:
after first publish, `e.writePos == 1`, so initial `e.gen` stays at zero until first time `e.writePos` wraps around zero